### ISSUE TITLE
Я идиот! убейте меня кто-нибудь!

### ISSUE THREAD

**Issue by a-x-** [bem/bem-forum-content-ru#909]
XJT/BEMS - Xtendable (almost)-json Transformation (BEMTREE killer) based on `Block Elem Mod State` concept

https://gist.github.com/a-x-/39242cb6ee00231900f0

**Comment by zxqfox** [comment]
Из сырых данных у тебя не получится при той же скорости иметь ту же гибкость, как ни крути.

При полностью декларативном подходе тебе нужно учесть все возможные кейсы, а это в принципе сделать невозможно, потому что их бесконечное кол-во и в каком-то виде нужен фолбек в императив.

Еще будет проблема с зависимостями по State — если хочешь добавить какую-то новую смысловую единицу — опиши как она будет работать со всем имеющимся. Это не так-то просто.

И еще лучше не завязываться на разделители ;-).

Попытка крутая, спасибо что зашарил.

p.s. Еще я не очень понял, почему ты не пытался смотреть в сторону всяких имеющихся трасляторов, типа http://stackoverflow.com/questions/1618038/xslt-equivalent-for-json и http://speranskydanil.github.io/json-translator/ — потому что они не знают про BEMS? наверное, да.

**Comment by a-x-** [comment]
Спасибо за рецензию!

Да, фолбэк предполагается. Это будет не json, а такой вырожденный js

Хочу, чтобы most used cases покрывались декларацией, а там где декларация могла бы стать очень _намагиченной_ не придумывать её, а использовать js

Хочется запилить такой JS(ON) конвертор JSON -> (BEM)JSON, совсем как XSLT, да
но в привязке к БЭМ(С)

Имеющиеся JSON: JSON -> JSON конверторы как есть не подойдут, но могут стать ядром имплементации.

p.s. XSLT, IMHO, страдает тем, что не адаптируется под декларативную XML и юзает вполне себе императивный синтаксис и получается это криво.

**Comment by zxqfox** [comment]
Про XSLT — есть такое. Как и во множестве других попыток перенести XSLT в js (коей является XJST, кстати говоря).

Но на другие конверторы все равно посмотри, так или иначе все завязываются на XSLT и пытаются перенести это в JS. Чужой опыт будет полезен как никогда.

**Comment by a-x-** [comment]
Буду смотреть :)

Про state буду думать,
я пока те кейсы обдумал, которые пришли в голову.

Например, на state нельзя матчиться в XJT, но можно в i-bem
Для обычного i-bem state может выглядеть неотличимо от mod

Есть потенциальная хрупкость разделения mod/state, попробую начать "конвертировать" проект Я.Картинок, тогда всё и "устаканится"

**Comment by zxqfox** [comment]
Да, большой проект сильно поможет понять проблемы в концепции и шлефануть реализацию. ;-)

**Comment by a-x-** [comment]
из json трансформаторов я очень люблю пользоваться jq,
он оказал влияние на запись доступа к данным 
.
.key
.key.nestedKey

**Comment by vithar** [comment]
Я так и не понял чем не подошёл BEMTREE? Можешь развернуто ответить?

**Comment by a-x-** [comment]
Привет, дам развёрнутый, но наверно сумбурный ответ.

__tl;dr__: Диссонанс: утверждается, что BEMTREE = декларативность,
но на деле нужно писать много кода на js ==> зашумленность, сложность
p.s. И в конце этих многобукв сказано про отдельную проблему на стыке priv <--> bem-xjst

См. [сравнение bemtree и xjt](https://gist.github.com/a-x-/39242cb6ee00231900f0#gistcomment-1717288). Допускаю, что этот вымышленный bemtree-файл не является типовым, буду рад посмотреть на такие примеры.

У нас в Картинках bemtree не используется (а можешь накидать репозиториев, где используется?), но из того что я видел, ситуация там схожая с bemhtml.
__Немного смещённой статистики по bemhtml__
Сейчас проверил, у нас например 286 bemhtml.js файлов,
224 (ок. 70%) из них используют функции.
Под сотню содержат логические операторы && ||
Такая же примерно картина и в СЕРПе.
[Поискал по яндекс коду](https://search.yandex-team.ru/codesearch?text=bemtree), нашёл некоторые сервисы, использующие bemtree, например wiki. У них немного xjst-файлов и в половине есть нехорошие признаки.
И это не от того, что у нас код кривой (хотя и не без этого).

__Немного примеров__:
Чтобы добавить модификатор нужно написать что-то типа
`def()(function(){this.mods = this.extend(this.mods || {}, {/*mods*/}); return applyNext()})`
возможно это из-за того, что считается, что модификаторы не нужно проставлять в шаблоне,
но тут, ИМХО, проблема в том, что модификаторы модификаторам рознь.

Миксы и многие другие моды нужно тоже оборачивать в функции, чтобы избежать проблем при доопределениях.

__Операции пере- и до- определения__ очень часты, и их приходится выполнять очень императивно.
__Вместо этого предлагаются__ _Merge strategies_ ([подробности](https://gist.github.com/a-x-/39242cb6ee00231900f0#gistcomment-1717317))

На каждом шагу API шаблонизатора подталкивает забыть про декларативность и написать js код

__Ещё__, т.к. BEMJSON позволяет описывать представление
почти все привы у нас описывают и маппинг данных и логику, и представление (почти как в PHP)
(XJT принимает не BEMJSON, а просто JSON что исключает скатывание в PHP)

Далее, __BEMTREE, это просто урезанный BEMHTML__, но на этом этапе хочется чего-то существенно отличающегося. Чего-то что позволяет решать задачи по переводу JSON в BEMJSON более лаконично.

__Границы между priv, bemtree, bemhtml размыты__ (обозначены на уровне соглашений, которые не исполняются).
Вот [пример ужасающего bemhtml](https://github.yandex-team.ru/mm-interfaces/fiji/blob/dev/./video/blocks-deskpad/series-table/__switcher/series-table__switcher.bemhtml.js)

----

p.s. __Ещё проблема__: блок имеющий логику/маппинг данных в priv не может быть встроен на уровне BEMTREE/BEMHTML,
только на стороне привов же его можно встроить в BEMJSON
__Предлагаемое решение__: регистрировать декларации блоков в неком хранилище в BEMPRIV, а вызываться они должны автоматически неявно на этапе XJT
В этих привах не должно быть BEMJSON'а, только маппинг данных и какая-нибудь хитрая логика.
